crates/turbopack-core/src/chunk/mod.rs

```diff
@@ -362,194 +367,305 @@ pub trait FromChunkableAsset: ChunkItem + Sized + Debug {
-                if !split
-                    && prev_chunk_items != chunk_items_count
-                    && chunk_items_count > 5000
-                    && prev_chunk_items > 1
-                {
-                    // Chunk is too large, cancel this algorithm and
-                    // restart with splitting from the start
-                    return Ok(None);
+                } else {
+                    return Ok(vec![ChunkContentGraphNode::ExternalAssetReference(
+                        reference,
+                    )]);
                 }
             }
         }
     }
 
+    if !new_processed_assets.is_empty() && new_chunk_items_count > 0 {
+        let mut state = state.lock();
+        state.processed_assets.extend(new_processed_assets);
+        state.chunk_items_count += new_chunk_items_count;
+    }
+
+    Ok(graph_nodes)
+}
+
+async fn chunk_item_to_graph_nodes<I>(
+    context: ChunkContentContext,
+    state: Arc<Mutex<ChunkContentState>>,
+    chunk_item: I,
+) -> Result<impl Iterator<Item = ChunkContentGraphNode<I>>>
+where
+    I: FromChunkableAsset + Eq + std::hash::Hash + Clone,
+{
+    Ok(chunk_item
+        .references()
+        .await?
+        .into_iter()
+        .map(|reference| reference_to_graph_nodes::<I>(context, Arc::clone(&state), *reference))
+        .try_join()
+        .await?
+        .into_iter()
+        .flatten())
+}
+
+/// The maximum number of chunk items that can be in a chunk before we split it
+/// into multiple chunks.
+const MAX_CHUNK_ITEMS_COUNT: usize = 5000;
+
+async fn chunk_content_internal_parallel<I>(
+    chunking_context: ChunkingContextVc,
+    entry: AssetVc,
+    additional_entries: Option<AssetsVc>,
+    split: bool,
+) -> Result<Option<ChunkContentResult<I>>>
+where
+    I: FromChunkableAsset + Eq + std::hash::Hash + Clone,
+{
+    let mut processed_assets: HashSet<(ChunkingType, AssetVc)> = HashSet::default();
+
+    // TODO(alexkirsz) We shouldn't need a clone here.
+    let additional_entries = if let Some(additional_entries) = additional_entries {
+        additional_entries.await?.clone_value().into_iter()
+    } else {
+        vec![].into_iter()
+    };
+
+    let entries = [entry].into_iter().chain(additional_entries);
+
+    // TODO(alexkirsz) We shouldn't need a clone here.
+    for entry in entries.clone() {
+        processed_assets.insert((ChunkingType::Placed, entry));
+    }
+
+    let root_graph_nodes = entries
+        .map(|entry| async move {
+            Ok(ChunkContentGraphNode::ChunkItem(
+                I::from_asset(chunking_context, entry).await?.unwrap(),
+            ))
+        })
+        .try_join()
+        .await?;
+
+    processed_assets.insert((ChunkingType::Placed, entry));
+
+    let context = ChunkContentContext {
+        chunking_context,
+        entry,
+        split,
+    };
+    let state = Arc::new(Mutex::new(ChunkContentState {
+        processed_assets,
+        chunk_items_count: 0,
+    }));
```

Comment:
Technically this should be doable without a Mutex (that's a bit expensive), by checking these constraints after handing a Future from FuturesUnordered. I guess this should be part of the trait.